src/bootstrap.coffee

```diff
@@ -34,27 +35,71 @@ loadPreloadedApps = ->
 	.catch (err) ->
 		utils.mixpanelTrack('Loading preloaded apps failed', { error: err })
 
+exchangeKey = ->
+	resinApi.get
+		resource: 'device'
+		options:
+			filter:
+				uuid: userConfig.uuid
+		customOptions:
+			apikey: userConfig.apiKey
+	.catchReturn([])
+	.timeout(config.apiTimeout)
+	.then ([ device ]) ->
+		if not device?
+			throw new ExchangeKeyError("Couldn't fetch device with provisioning key")
+		# We found the device, we can try to generate a working device key for it
+		request.postAsync("#{config.apiEndpoint}/api-key/device/#{device.id}/device-key")
+		.spread (res, body) ->
+			if res.status != 200
+				throw new ExchangeKeyError("Couldn't generate device key with provisioning key")
+			userConfig.deviceApiKey = body
+		.return(device)
+
 bootstrap = ->
 	Promise.try ->
 		userConfig.deviceType ?= 'raspberry-pi'
 		if userConfig.registered_at?
 			return userConfig
-		deviceRegister.register(resinApi, userConfig)
+
+		deviceRegister.register(
+			userId: userConfig.userId
+			applicationId: userConfig.applicationId
+			uuid: userConfig.uuid
+			deviceType: userConfig.deviceType
+			deviceApiKey: userConfig.deviceApiKey
+			provisioningApiKey: userConfig.apiKey
+			apiEndpoint: config.apiEndpoint
+		)
 		.timeout(config.apiTimeout)
 		.catch DuplicateUuidError, ->
+			console.log('Hit duplcate UUID')
```

Comment:
Initially, but I then expanded on it so it's possible to see in the supervisor logs what's going on with registration - I'll improve this one's message though